### DISCUSSION TITLE
Performance measurements: Parsing dates

### DISCUSSION THREAD

**Discussion by westnordost** [streetcomplete/StreetComplete#2740] [General]
[Recently I mentioned](https://github.com/streetcomplete/StreetComplete/issues/2506#issuecomment-792370165) that parsing dates in Java is quite slow and that 50% of the total parsing time of OpenStreetMap map data is spent on parsing the dates.

I advertised to do this bigger refactor https://github.com/streetcomplete/StreetComplete/pull/2678 because then this would be a good opportunity to get rid of the `SimpleDateFormat` parsing and use the more performant `kotlinx-datetime`, which is based on the Java 8 time API, also available on Android (if certain build configurations are made).

So, I now did a measurement to see how slow or fast exactly each method is. Here it is:

<table>
<tr>
<th>Parse method</th><th colspan="3">Average time to parse one date (on one core)</th>
</tr>
<tr>
<td></td><td>Samsung S10e<br>Android 10<br>(2019)</td><td>Samsung S4 Mini Plus<br>Android 9<br>(2015)</td><td>Samsung S4 Mini<br>Android 9<br>(2013)</td><td>Sony Xperia Z1<br>Android 5.1<br>(2013)</td>
</tr>
<tr>
<td>
Naïve (and wrong), using substring (no validation)
</td>
<td>1 μs</td>
<td>1 μs</td>
<td>1 μs</td>
<td>5 μs</td>
</tr>
<tr>
<td>
Naïve (and wrong), using a regex
</td>
<td>8 μs</td>
<td>39 μs</td>
<td>51 μs</td>
<td>54 μs</td>
</tr>
<tr>
<td>
<code>GregorianCalendar</code>, using substring (no validation)
</td>
<td>2 μs</td>
<td>11 μs</td>
<td>13 μs</td>
<td>26 μs</td>
</tr>
<tr>
<tr>
<td>
<code>GregorianCalendar</code>, using a regex
</td>
<td>11 μs</td>
<td>56 μs</td>
<td>66 μs</td>
<td>77 μs</td>
</tr>
<tr>
<td>
<code>SimpleDateFormat</code>
</td>
<td>12 μs</td>
<td>53 μs</td>
<td>62 μs</td>
<td>180 μs</td>
</tr>
<tr>
<td>
<code>LocalDateTime.parse</code>
</td>
<td>4 μs</td>
<td>28 μs</td>
<td>32 μs</td>
<td>48 μs</td>
</tr>
<tr>
<td>
<code>Instant.parse</code>
</td>
<td>6 μs</td>
<td>42 μs</td>
<td>55 μs</td>
<td>77 μs</td>
</tr>
</table>

And yeah, even 180 μs appears little, but let's parse 30000 OpenStreetMap elements which each have a date attached and we are at 5.4 seconds for the date parsing alone.

![z1](https://user-images.githubusercontent.com/4661658/114462516-6e93e600-9be3-11eb-91a4-e524aa567fe1.png)
![image](https://user-images.githubusercontent.com/4661658/114864165-f69b0b00-9df0-11eb-90ed-eb56efdbb20e.png)
![image](https://user-images.githubusercontent.com/4661658/114864279-17636080-9df1-11eb-9b4f-569dfe014197.png)
![s10e](https://user-images.githubusercontent.com/4661658/114462517-6f2c7c80-9be3-11eb-8304-5f72b16ce80e.png)

That old 5-7 years old mid-range phones are significantly slower than a relatively recent high-end phone is no surprise.

The naïve (and wrong) parsing that simply splits up the date string is each the bottom-line, it is pretty much not possible to make it any faster than that. 
There are two observations:
- The new Java 8 time API (`Instant`) is 20% to 100% faster, depending on the phone / Android version. `SimpleDateFormat` (old Java time API) is consistently the slowest.
- On old Android versions (below Android 8), the difference seems to be greater. This leads me to believe that CPU speed is not the only factor here but the Android version as well: The new Java 8 time API is available on Android since API level 26 (Android 8). I find it probable that Google could have backported some performance improvements from the new time API to the old time API (`Date`, `SimpleDateFormat`). Could have other reaons of course.

# Conclusion

Migrating to `kotlinx-datetime` which uses the Java 8 time API (using the `Instant.parse` method) will reduce date parsing time by 10%-50%, for older devices possibly more.

This then will result in a decrease of download times by up to 25%. So, a download that took 20 seconds before will then take just 14 seconds. It will not really have any effect on anything else because that is the only place in the app where dates are parsed (en mass).

Not to mention, this is something that needs to be done anyway if one day an iOS version was made (#1892).

# Code

<details>
<summary>Code used for measuring this</summary>

```kotlin

package de.westnordost.streetcomplete

import org.junit.Test
import java.text.SimpleDateFormat
import java.time.Instant
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.math.roundToLong
import kotlin.system.measureTimeMillis

class DateParsingPerformanceTest {

    @Test fun dateParsingPerformance() {
        val date = "2015-05-07T09:58:16Z"

        val dateRegex = Regex("(\\d{4})-(\\d{2})-(\\d{2})T(\\d{2}):(\\d{2}):(\\d{2})Z")

        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'")
        simpleDateFormat.timeZone = TimeZone.getTimeZone("UTC")

        val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")

        println("Parsing date: Naïve (and wrong), using substring (no validation) - ${measure {
            val year = date.substring(0,4).toInt()
            val month = date.substring(5,7).toInt()
            val day = date.substring(8,10).toInt()
            val hour = date.substring(11,13).toInt()
            val min = date.substring(14,16).toInt()
            val sec = date.substring(17,18).toInt()
            year*365*(month*31*(day*24*(hour*60*(min*60*(sec*1000)))))
        }} μs")

        println("Parsing date: Naïve (and wrong), using a regex - ${measure {
            val values = dateRegex.matchEntire(date)!!.groupValues
            val year = values[1].toInt()
            val month = values[2].toInt()
            val day = values[3].toInt()
            val hour = values[4].toInt()
            val min = values[5].toInt()
            val sec = values[6].toInt()
            year*365*(month*31*(day*24*(hour*60*(min*60*(sec*1000)))))
        }} μs")

        println("Parsing date: GregorianCalendar, using substring (no validation) - ${measure {
            val year = date.substring(0,4).toInt()
            val month = date.substring(5,7).toInt()
            val day = date.substring(8,10).toInt()
            val hour = date.substring(11,13).toInt()
            val min = date.substring(14,16).toInt()
            val sec = date.substring(17,18).toInt()
            GregorianCalendar(year, month - 1, day, hour, min, sec).time
        }} μs")

        println("Parsing date: GregorianCalendar, using a regex - ${measure {
            val values = dateRegex.matchEntire(date)!!.groupValues
            val year = values[1].toInt()
            val month = values[2].toInt()
            val day = values[3].toInt()
            val hour = values[4].toInt()
            val min = values[5].toInt()
            val sec = values[6].toInt()
            GregorianCalendar(year, month - 1, day, hour, min, sec).time
        }} μs")

        println("Parsing date: SimpleDateFormat - ${measure {
            simpleDateFormat.parse(date)
        }} μs")

        println("Parsing date: LocalDateTime.parse - ${measure {
            LocalDateTime.parse(date, dateTimeFormatter)
        }} μs")

        println("Parsing date: Instant.parse - ${measure {
            Instant.parse(date)
        }} μs")
    }
}
/** Measure how long the block needs for execution on average in microseconds, rounded */
private inline fun measure(block: () -> Unit): Long {
    val iterations = 1_000_000
    val millis = measureTimeMillis { repeat(iterations) { block() } }
    return ((10000 * millis / iterations).toDouble() / 10).roundToLong()
}

```
</details>

**Comment by smichel17** [comment]
> From that bottom line, we see that the Android 5 phone is 5 times slower than the Android 11 one.

I think you've mixed up the numbers somewhere, because in the graphs, the 2019 phone is slower than the 2013 phone.

**Reply by smichel17** [reply]
Ack, never mind, inconsistent axes strike again!

**Comment by Helium314** [comment]
Some questions that make me feel a bit stupid...
* Where is date parsing required? Apparently it is necessary for every element, so I guess it is the modification time... but why would this be stored as string and not as a simple (int or long) timestamp?
* Where do I actually put the code if I want to run it on my phone?(I'm curious how a weaker CPU than Xperia Z1, but running Android 9 affects this.)

**Reply by matkoniecz** [reply]
> Where is date parsing required

For example "is this object still existing" and "is this road construction complete" quests. I am pretty sure that it is not done for every element.

> Where do I actually put the code if I want to run it on my phone?

Add it to tests, run this specific test. (in general Android Studio is used)

**Reply by westnordost** [reply]
The timestamp IS saved as a long (Unix timestamp). The issue is that the OSM Api transmits this as a "yyyy-mm-dd'T'hh:mm:ss'Z'" string, so that needs to be parsed.

You can put the Code anywhere in the AndroidTest folder, then rightclick on it and run it as a test (not in debug, that will be slower)

**Reply by westnordost** [reply]
I'd also be curious on how it performs on new but slow phones. Did you try it out, @Helium314 ? I could include your measurements in the above table

**Reply by Helium314** [reply]
Not yet, it will take some longer time before I have physical connection to the PC running Android Studio...
But if you would like to know, I'll try to create some mini-app containing the code.

**Reply by westnordost** [reply]
How, if you have no access to an Android Studio?

**Reply by Helium314** [reply]
Remote access is working, but as I understand it for running these test I need to plug my phone to have adb connection.

**Reply by Helium314** [reply]
So... finally results on my phones running Lineage 16.0:
S4 Mini has a Snapdragon 400 (1.7 GHz, dual core, 32 bit)
S4 Mini Plus has a Snapdragon 410 (1.2 GHz, quad core, 64 bit but Lineage only working as 32 bit version)

method | S4 Mini, 1 core | S4 Mini, 2 cores | S4 Mini Plus, 1 core | S4 Mini Plus, 4 cores
--|--|--|--|--
Naïve (and wrong), using substring (no validation) | 1 μs | 1 μs | 2 µs | 1 µs
Naïve (and wrong), using a regex | 63 μs | 51 μs | 61 µs | 39 µs
GregorianCalendar, using substring (no validation) | 25 μs | 13 μs | 20 µs | 11 µs
GregorianCalendar, using a regex | 94 μs | 66 μs | 80 µs | 56 µs
SimpleDateFormat | 82 µs | 62 μs | 67 µs | 53 µs
LocalDateTime.parse | 33 μs | 32 μs | 30 µs | 28 µs
Instant.parse | 55 μs | 55 μs | 46 µs | 42 µs

`Naïve (and wrong), using substring (no validation)` sometimes takes significantly longer. On both phones I ran the tests multiple times and each had one run where this method took 6 or 7 µs.
Other tests are reproducible within ca 10%.

If someone want to run the test, simply extract and install the attached [app](https://github.com/streetcomplete/StreetComplete/files/6316534/date_parse_test.zip). Start app, click `start test` and wait (up to) a few minutes (nothing else in this app is actually working).
The last two tests require Android 8 and will not appear on older versions.

**Comment by westnordost** [comment]
How did you let it run on multiple cores?

On April 15, 2021 11:28:06 AM GMT+02:00, Helium314 ***@***.***> wrote:
>So... finally results on my phones running Lineage 16.0:>
>S4 Mini has a Snapdragon 400 (1.7 GHz, dual core)>
>S4 Mini Plus has a Snapdragon 410 (1.2 GHz, quad core)>
>>
>method | S4 Mini, 1 core | S4 Mini, 2 cores | S4 Mini Plus, 1 core | S4
>Mini Plus, 4 cores>
>--|--|--|--|-->
>Naïve (and wrong), using substring (no validation) | 1 μs | 1 μs | 2 µs
>| 1 µs>
>Naïve (and wrong), using a regex | 63 μs | 51 μs | 61 µs | 39 µs>
>GregorianCalendar, using substring (no validation) | 25 μs | 13 μs | 20
>µs | 11 µs>
>GregorianCalendar, using a regex | 94 μs | 66 μs | 80 µs | 56 µs>
>SimpleDateFormat | 82 µs | 62 μs | 67 µs | 53 µs>
>LocalDateTime.parse | 33 μs | 32 μs | 30 µs | 28 µs>
>Instant.parse | 55 μs | 55 μs | 46 µs | 42 µs>
>>
>`Naïve (and wrong), using substring (no validation)` sometimes takes
>significantly longer. On both phones I ran the tests multiple times and
>each had one run where this method took 6 or 7 µs.>
>Other tests are reproducible within ca 10%.>
>>
>If someone want to run the test, simply extract and install the
>attached
>[app](https://github.com/streetcomplete/StreetComplete/files/6316534/date_parse_test.zip).
>Start app, click `start test` and wait (up to) a few minutes (nothing
>else in this app is actually working).>
>The last two tests require Android 8 and will not appear on older
>versions.>
>>
>-- >
>You are receiving this because you authored the thread.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/streetcomplete/StreetComplete/discussions/2740#discussioncomment-614375

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Reply by Helium314** [reply]
I simply disable all cores except one (on the phone, not in the app) if I want to do a single core test.
So no specific code for using multiple cores, this is done by Android.

**Reply by westnordost** [reply]
Oh ok, the code actually runs only on one core in any case. If there are differences it's because other processes do stuff too.

Anyway, the test has disappointing results, only 12-26% faster. I wonder why this is so different.

**Reply by westnordost** [reply]
Do the tests run on the ui thread?

**Reply by Helium314** [reply]
Yes. I disabled WiFI and most other apps so there would be little to do in the background.

**Reply by westnordost** [reply]
Hmm, maybe if something runs too long on the main thread, android suspends the app (more). Could you try if it has different results if the tests run in background?

Also, 4,3" is a very smal screen. The bottom line was 4,7". Are there any display problems on that screen size?

**Reply by Helium314** [reply]
Starting and switching to home screen makes no difference.
When switching the screen off the app definitely get suspended and tests easily take 10 times longer.

I have an overlay showing current CPU frequency. When using only one core the frequency is always at maximum, so I think the tests are not suspended (there is no other app doing significant work in the background).
With more cores, only one core is at maximum freqency all the time, the others change (confirming that the tests are mostly using only one core).

> Also, 4,3" is a very smal screen. The bottom line was 4,7". Are there any display problems on that screen size?

No problems.
A larger display would be nice, but apparently my requirements are too strict... small phone size (< 65 mm) and OLED display is a very neglected category; the iPhone 12 mini was the only new device here in the past 5 years.

**Reply by westnordost** [reply]
Well I meant, in a background thread while the app is in foreground. Not as a service or something, just that it does not block the UI thread.

Small device: Yeah I am also a small device lover. This is why I got the S10e (bought it second hand at a lower price a new A41 would cost). It's 5.8" though, that's the smallest it gets nowadays and already too large to conveniently use with one hand. I recently was amazed by some videos of mini-phones with a display that is even smaller than... really old phones - 3 inch. That's about the size of a credit card: [Palm Pepito](https://www.youtube.com/watch?v=34qha-0CTrk) (cool UI, tiny battery) and [Jelly 2](https://www.youtube.com/watch?v=-t98Vqki3Qk) (ok battery). Though, I am really curious about these two, they are probably an inch too small for my liking. It would be interesting to add support for such a small display size though if anyone really used these.

**Reply by Helium314** [reply]
> Well I meant, in a background thread while the app is in foreground. Not as a service or something, just that it does not block the UI thread.

I didn't manage to make it work... what do I need to do?

> Small device:

The tiny phones look quite nice, and seem fast enough. I especially like the IR blaster on the Jelly 2 (using it almost daily on my S4 mini).
But something between those 3" and the common 6" phones would probably be more suited for (my) normal use.

**Reply by westnordost** [reply]
Do you use Kotlin and the code is in a MainActivity? Then easiest would be

```kotlin
thread {
    val results = runTheTests();
    // anything touching the view must execute on the UI thread in Android
    runOnUiThread {
        displayTheResults(results);
    }
}
```

**Reply by Helium314** [reply]
Thanks! It's in a fragment, but `activity?.runOnUiThread` was all that needed to be changed.

When running the phone on a single core this is 20-30% slower, except for both `no validation` tests.
With all cores active, the last tests are a bit faster (ran the tests twice to confirm)
method | main thread | background thread
-- | -- | -- 
SimpleDateFormat | 53 µs | 51 µs
LocalDateTime.parse | 28 µs | 27 µs
Instant.parse | 42 µs | 39 µs


Is a valid timestamp actually required for all elements? If not it could be feasible to exclude some elements from date parsing...

**Reply by westnordost** [reply]
Ok, so not really any difference then. Thanks!

Yes, a valid timestamp is required for all elements because the downloader does not know how the data will be used.

**Reply by smichel17** [reply]
Are they serialized before processing, or analyzed while still in memory, and then persisted? If the latter, maybe the date parsing could be done lazily.

**Reply by westnordost** [reply]
Already considered this
